### PR TITLE
Volume transmit api

### PR DESCRIPTION
**Pull request by heavenlyhash**
Add volume snapshots and (inter-host!) transmission to the host API.

### CODE REVIEW

File: host/host.go [hunk 1]
```diff
@@ -264,6 +259,16 @@ func runDaemon(args *docopt.Args) {
 	}
 	shutdown.BeforeExit(func() { hb.Close() })
 
+	cluster, err := cluster.NewClient()
+	if err != nil {
+		shutdown.Fatal(err)
+	}
+
+	router, err := serveHTTP(&Host{state: state, backend: backend}, &attachHandler{state: state, backend: backend}, cluster, vman)
```
Comment (titanous) [inline host/host.go]: This should be split up so that each field is on it's own line.

File: host/volume/api/http.go [hunk 2]
```diff
@@ -26,6 +31,10 @@ func (api *HTTPAPI) RegisterRoutes(r *httprouter.Router) {
 	r.GET("/storage/volumes", api.List)
 	r.GET("/storage/volumes/:volume_id", api.Inspect)
 	r.PUT("/storage/volumes/:volume_id/snapshot", api.Snapshot)
+	// takes host and volID parameters, triggers a send on the remote host and give it a list of snaps already here, and pipes it into recv
+	r.POST("/storage/volumes/:volume_id/pullSnapshot", api.Pull)
+	// responds with a snapshot stream binary.  only works on snapshots, takes 'haves' parameters, usually called by a node that's servicing a 'pullSnapshot' request
+	r.POST("/storage/volumes/:volume_id/send", api.Send)
```
Comment (titanous) [inline host/volume/api/http.go]: I think a GET with a specific media type would make sense here.

File: host/volume/api/http.go [hunk 3]
```diff
@@ -26,6 +31,10 @@ func (api *HTTPAPI) RegisterRoutes(r *httprouter.Router) {
 	r.GET("/storage/volumes", api.List)
 	r.GET("/storage/volumes/:volume_id", api.Inspect)
 	r.PUT("/storage/volumes/:volume_id/snapshot", api.Snapshot)
+	// takes host and volID parameters, triggers a send on the remote host and give it a list of snaps already here, and pipes it into recv
+	r.POST("/storage/volumes/:volume_id/pullSnapshot", api.Pull)
```
Comment (titanous) [inline host/volume/api/http.go]: `pull_snapshot` is a bit better. I'm not a huge fan of the current route, but I don't have any better suggestions.

File: host/volume/api/http.go [hunk 4]
```diff
@@ -110,5 +119,83 @@ func (api *HTTPAPI) Inspect(w http.ResponseWriter, r *http.Request, ps httproute
...
+	}
+
+	httphelper.JSON(w, 200, snap.Info())
+}
+
+func (api *HTTPAPI) Pull(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	volumeID := ps.ByName("volume_id")
+
+	pull := &volume.PullCoordinate{}
+	if err := json.NewDecoder(r.Body).Decode(&pull); err != nil {
```
Comment (titanous) [inline host/volume/api/http.go]: Use `httphelper.DecodeJSON`

File: host/volume/api/http.go [hunk 5]
```diff
@@ -110,5 +119,83 @@ func (api *HTTPAPI) Inspect(w http.ResponseWriter, r *http.Request, ps httproute
...
+		return
+	}
+
+	httphelper.JSON(w, 200, snap.Info())
+}
+
+func (api *HTTPAPI) Send(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	volumeID := ps.ByName("volume_id")
+
+	haves := make([]json.RawMessage, 0)
```
Comment (titanous) [inline host/volume/api/http.go]:
```go
var haves []json.RawMessage
```

File: host/volume/api/http.go [hunk 6]
```diff
@@ -110,5 +119,83 @@ func (api *HTTPAPI) Inspect(w http.ResponseWriter, r *http.Request, ps httproute
...
+	if err := json.NewDecoder(r.Body).Decode(&haves); err != nil {
```
Comment (titanous) [inline host/volume/api/http.go]: Use `httphelper.DecodeJSON`

File: host/volume/api/http.go [hunk 7]
```diff
@@ -110,5 +119,83 @@ func (api *HTTPAPI) Inspect(w http.ResponseWriter, r *http.Request, ps httproute
...
+		httphelper.Error(w, err)
+		return
+	}
+
+	// TODO proxy this writer...? poor error bubbling here
```
Comment (titanous) [inline host/volume/api/http.go]: Is this actually an issue?

File: host/volume/zfs/zfs.go [hunk 8]
```diff
@@ -161,6 +184,208 @@ func (b *Provider) DestroyVolume(vol volume.Volume) error {
...
+func (b *Provider) SendSnapshot(vol volume.Volume, haves []json.RawMessage, output io.Writer) error {
+	zvol, err := b.owns(vol)
+	if err != nil {
+		return err
+	}
+	if !vol.IsSnapshot() {
+		return fmt.Errorf("can only send a snapshot")
+	}
+	// zfs recv can only really accept snapshots that apply to the current tip
+	latestRemote := ""
```
Comment (titanous) [inline host/volume/zfs/zfs.go]:
```go
var latestRemote string
```

File: host/volume/zfs/zfs.go [hunk 9]
```diff
@@ -161,6 +184,208 @@ func (b *Provider) DestroyVolume(vol volume.Volume) error {
...
+	if haves != nil && len(haves) > 0 {
+		have := &zfsHaves{}
+		if err := json.Unmarshal(haves[len(haves)-1], have); err == nil {
+			latestRemote = have.SnapID
+		}
+	}
+	// look for intersection of existing snapshots on this volume; if so do incremental
+	parentName := strings.Split(zvol.dataset.Name, "@")[0]
```
Comment (titanous) [inline host/volume/zfs/zfs.go]: Use `SplitN`

File: host/volume/zfs/zfs.go [hunk 10]
```diff
@@ -161,6 +184,208 @@ func (b *Provider) DestroyVolume(vol volume.Volume) error {
...
+	}
+	snapshots, err := parentDataset.Snapshots()
+	if err != nil {
+		return err
+	}
+	// we can fly incremental iff the latest snap on the remote is available here
+	useIncremental := false
+	if latestRemote != "" {
+		for _, snap := range snapshots {
+			if strings.Split(snap.Name, "@")[1] == latestRemote {
```
Comment (titanous) [inline host/volume/zfs/zfs.go]: Use `SplitN`

File: pkg/cluster/host.go [hunk 11]
```diff
@@ -97,6 +108,30 @@ func (c *hostClient) CreateVolume(providerId string) (*volume.Info, error) {
...
+	var res volume.Info
+	pull := volume.PullCoordinate{
+		HostID:     sourceHostID,
+		SnapshotID: sourceSnapID,
+	}
+	err := c.c.Post(fmt.Sprintf("/storage/volumes/%s/pullSnapshot", receiveVolID), pull, &res)
+	return &res, err
+}
+
+func (c *hostClient) SendSnapshot(snapID string, assumeHaves []json.RawMessage) (io.ReadCloser, error) {
```
Comment (titanous) [inline pkg/cluster/host.go]: Where do I get `assumeHaves`?

Comment (titanous) [inline pkg/cluster/host.go]: Ah, nvm.

**Comment by titanous** [comment]
Overall this looks good. My comments are pretty minor. Looking forward to getting this merged!

File: host/volume/api/http.go (hunk 6 above)
Comment (lfaraone) [inline host/volume/api/http.go]: What does that method actually help with?

File: host/volume/api/http.go (hunk 2 above)
Comment (heavenlyhash) [inline host/volume/api/http.go]: It does feel like a GET operation, but there's a request body to hold the "haves".

`Content-Type: application/zfs-snapshot-stream`?  (I haven't found any other examples in the wild yet.)

File: host/volume/api/http.go (hunk 6 above)
Comment (archSeer) [inline host/volume/api/http.go]: Yeah I agree, I think we should get rid of `httphelper.DecodeJSON` altogether.

File: host/volume/api/http.go (hunk 7 above)
Comment (heavenlyhash) [inline host/volume/api/http.go]: Yeah, dropping todo.  It's not a very great story if the send command gets an error after starting the output stream, but there's not much for ways to backtrack on an http 200 that's already been flushed.

File: host/volume/api/http.go (hunk 2 above)
Comment (titanous) [inline host/volume/api/http.go]: `application/vnd.zfs.snapshot-stream` would be appropriate.

Comment (titanous) [inline host/volume/api/http.go]: Oh, and GET requests can have bodies.

File: host/volume/api/http.go (hunk 6 above)
Comment (titanous) [inline host/volume/api/http.go]: DecodeJSON can/will be expanded to do things like better error handling, debug logging, and API version adjustments. I don't think removing it makes sense.

**Comment by titanous** [comment]
LGTM

**Pull request merged**